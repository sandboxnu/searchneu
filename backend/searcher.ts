/*
 * This file is part of Search NEU and licensed under AGPL3.
 * See the license file in the root folder for details.
 */

import _ from 'lodash';
import { Course, Section } from '@prisma/client';
import prisma from './prisma';
import elastic, { Elastic } from './elastic';
import HydrateSerializer from './database/serializers/hydrateSerializer';
import HydrateCourseSerializer from './database/serializers/hydrateCourseSerializer';
import macros from './macros';
import {
  EsQuery, QueryNode, ExistsQuery, TermsQuery, TermQuery, LeafQuery, MATCH_ALL_QUERY, RangeQuery,
  EsFilterStruct, EsAggFilterStruct, FilterInput, FilterPrelude, AggFilterPrelude, SortInfo, Range,
  SearchResults, SingleSearchResult, PartialResults, EsResultBody, EsMultiResult, AggResults,
} from './search_types';

type CourseWithSections = Course & { sections: Section[] };

class Searcher {
  elastic: Elastic;

  subjects: Set<string>;

  filters: FilterPrelude;

  aggFilters: AggFilterPrelude;

  AGG_RES_SIZE: number;

  COURSE_CODE_PATTERN: RegExp;

  constructor() {
    this.elastic = elastic;
    this.subjects = null;
    this.filters = Searcher.generateFilters();
    this.aggFilters = _.pickBy<EsFilterStruct, EsAggFilterStruct>(this.filters, (f): f is EsAggFilterStruct => f.agg !== false);
    this.AGG_RES_SIZE = 1000;
    this.COURSE_CODE_PATTERN = /^\s*([a-zA-Z]{2,4})\s*(\d{4})?\s*$/i;
  }

  static generateFilters(): FilterPrelude {
    // type validating functions
    const isString = (arg: any): arg is string => {
      return typeof arg === 'string';
    };

    const isStringArray = (arg: any): arg is string[] => {
      return Array.isArray(arg) && arg.every((elem) => isString(elem));
    };

    const isTrue = (arg: any): arg is true => {
      return typeof arg === 'boolean' && arg;
    };

    const isNum = (arg: any): arg is number => {
      return typeof arg === 'number';
    };

    const isRange = (arg: any): arg is Range => {
      return _.difference(Object.keys(arg), ['min', 'max']).length === 0 && isNum(arg.min) && isNum(arg.max);
    };

    // filter-generating functions
    const getSectionsAvailableFilter = (): ExistsQuery => {
      return { exists: { field: 'sections' } };
    };

    const getNUpathFilter = (selectedNUpaths: string[]): TermsQuery => {
      return { terms: { 'class.nupath.keyword': selectedNUpaths } };
    };

    const getSubjectFilter = (selectedSubjects: string[]): TermsQuery => {
      return { terms: { 'class.subject.keyword': selectedSubjects } };
    };

    // note that { online: false } is never in filters
    const getOnlineFilter = (selectedOnlineOption: boolean): TermQuery => {
      return { term: { 'sections.online': selectedOnlineOption } };
    };

    const getClassTypeFilter = (selectedClassTypes: string[]): TermsQuery => {
      return { terms: { 'sections.classType.keyword': selectedClassTypes } };
    };

    const getTermIdFilter = (selectedTermId: string): TermQuery => {
      return { term: { 'class.termId': selectedTermId } };
    };

    const getRangeFilter = (selectedRange: Range): RangeQuery => {
      return { range: { 'class.classId': { gte: selectedRange.min, lte: selectedRange.max } } };
    };

    return {
      nupath: { validate: isStringArray, create: getNUpathFilter, agg: 'class.nupath.keyword' },
      subject: { validate: isStringArray, create: getSubjectFilter, agg: 'class.subject.keyword' },
      online: { validate: isTrue, create: getOnlineFilter, agg: false },
      classType: { validate: isStringArray, create: getClassTypeFilter, agg: 'sections.classType.keyword' },
      sectionsAvailable: { validate: isTrue, create: getSectionsAvailableFilter, agg: false },
      classIdRange: { validate: isRange, create: getRangeFilter, agg: false },
      termId: { validate: isString, create: getTermIdFilter, agg: false },
    };
  }

  async initializeSubjects(): Promise<void> {
    if (!this.subjects) {
      this.subjects = new Set((await prisma.course.findMany({ select: { subject: true }, distinct: ['subject'] })).map((obj) => obj.subject));
    }
  }

  /**
   * return a set of all existing subjects of classes
   */
  getSubjects(): Set<string> {
    return this.subjects;
  }

  /**
   * Remove any invalid filter with the following criteria:
   * 1. Correct key string and value type;
   * 2. Check that { online: false } should never be in filters
   *
   * A sample filters JSON object has the following format:
   * { 'nupath': string[],
   *   'college': string[],
   *   'subject': string[],
   *   'online': boolean,
   *   'classType': string }
   *
   * @param {object} filters The json object represting all filters on classes
   */
  validateFilters(filters: FilterInput): FilterInput {
    const validFilters: FilterInput = {};
    Object.keys(filters).forEach((currFilter) => {
      if (!(currFilter in this.filters)) {
        macros.log('Invalid filter key.', currFilter);
      } else if (!(this.filters[currFilter].validate(filters[currFilter]))) {
        macros.log('Invalid filter value type.', currFilter);
      } else {
        validFilters[currFilter] = filters[currFilter];
      }
    });
    return validFilters;
  }

  getFields(): string[] {
    return [
      'class.name^2', // Boost by 2
      'class.name.autocomplete',
      'class.subject^4',
      'class.classId^3',
      'sections.profs',
      'sections.crn',
      'employee.name^2',
      'employee.emails',
      'employee.phone',
    ];
  }

  /**
   * Get elasticsearch query
   */
  generateQuery(query: string, termId: string, userFilters: FilterInput, min: number, max: number, aggregation: string = ''): EsQuery {
    const fields: string[] = this.getFields();
    // text query from the main search box
    const matchTextQuery: LeafQuery = query.length > 0
      ? {
        multi_match: {
          query: query,
          type: 'most_fields', // More fields match => higher score
          fields: fields,
        },
      }
      : MATCH_ALL_QUERY;

    // use lower classId has tiebreaker after relevance
    const sortByClassId: SortInfo = { 'class.classId.keyword': { order: 'asc', unmapped_type: 'keyword' } };

    // filter by type employee
    const isEmployee: LeafQuery = { term: { type: 'employee' } };
    const areFiltersApplied: boolean = Object.keys(userFilters).length > 0;
    const requiredFilters: FilterInput = { termId: termId, sectionsAvailable: true };
    const filters: FilterInput = { ...requiredFilters, ...userFilters };

    const classFilters: QueryNode[] = _(filters).pick(Object.keys(this.filters)).toPairs().map(([key, val]) => this.filters[key].create(val))
      .value();

    const aggQuery = !aggregation ? undefined : {
      [aggregation]: {
        terms: { field: this.aggFilters[aggregation].agg, size: this.AGG_RES_SIZE },
      },
    };

    // compound query for text query and filters
    return {
      from: min,
      size: max - min,
      sort: ['_score', sortByClassId],
      query: {
        bool: {
          must: matchTextQuery,
          filter: {
            bool: {
              should: [
                { bool: { must: classFilters } },
                ...(!areFiltersApplied) ? [isEmployee] : [],
              ],
            },
          },
        },
      },
      aggregations: aggQuery,
    };
  }

  generateMQuery(query: string, termId: string, min: number, max: number, filters: FilterInput): EsQuery[] {
    const validFilters: FilterInput = this.validateFilters(filters);

    const queries: EsQuery[] = [this.generateQuery(query, termId, validFilters, min, max)];

    for (const fKey of Object.keys(this.aggFilters)) {
      const everyOtherFilter: FilterInput = _.omit(filters, fKey);
      queries.push((this.generateQuery(query, termId, everyOtherFilter, 0, 0, fKey)));
    }
    return queries;
  }

  async getSearchResults(query: string, termId: string, min: number, max: number, filters: FilterInput): Promise<PartialResults> {
    const queries = this.generateMQuery(query, termId, min, max, filters);
    const results: EsMultiResult = await elastic.mquery(`${elastic.CLASS_INDEX},${elastic.EMPLOYEE_INDEX}`, queries);
    return this.parseResults(results.body.responses, Object.keys(this.aggFilters));
  }

  parseResults(results: EsResultBody[], filters: string[]): PartialResults {
    return {
      output: results[0].hits.hits,
      resultCount: results[0].hits.total.value,
      took: results[0].took,
      aggregations: _.fromPairs(filters.map((filter, idx) => {
        return [filter, results[idx + 1].aggregations[filter].buckets.map((aggVal) => { return { value: aggVal.key, count: aggVal.doc_count } })];
      })),
    };
  }

  async getOneSearchResult(subject: string, classId: string, termId: string) : Promise<SingleSearchResult> {
    const start = Date.now();
    const result = await prisma.course.findOne({ where: { uniqueCourseProps: { classId, subject, termId } }, include: { sections: true } });
    const serializer = new HydrateCourseSerializer();
    let results = result ? await serializer.bulkSerialize([result]) : {};
    results = Object.values(results); // necessary because results looks like { some_class_id: { class: { ... }} and we want [{ class: { ...}}]
    return {
      results,
      resultCount: result === null ? 0 : 1,
      took: 0,
      hydrateDuration: Date.now() - start,
      aggregations: result ? this.getSingleResultAggs(result) : { nupath: [], subject: [], classType: [] },
    };
  }

  getSingleResultAggs(result: CourseWithSections) : AggResults {
    return {
      nupath: result.nupath.map((val) => { return { value: val, count: 1 } }),
      subject: [{ value: result.subject, count: 1 }],
      classType: [{ value: result.sections[0].classType, count: 1 }],
    };
  }

  /**
   * Search for classes and employees
   * @param  {string}  query  The search to query for
   * @param  {string}  termId The termId to look within
   * @param  {integer} min    The index of first document to retreive
   * @param  {integer} max    The index of last document to retreive
   */
  async search(query: string, termId: string, min: number, max: number, filters: FilterInput = {}): Promise<SearchResults> {
    await this.initializeSubjects();
    const start = Date.now();
    let results; let resultCount; let hydrateDuration; let took; let aggregations;

    // if we know that the query is of the format of a course code, we want to return only one result
    const patternResults = query.match(this.COURSE_CODE_PATTERN);
    const subject = patternResults ? patternResults[1].toUpperCase() : '';
    if (patternResults && macros.isNumeric(patternResults[2]) && (this.getSubjects()).has(subject)) {
      ({
        results, resultCount, took, hydrateDuration, aggregations,
      } = await this.getOneSearchResult(subject, patternResults[2], termId));
    } else {
      const searchResults = await this.getSearchResults(query, termId, min, max, filters);
      ({ resultCount, took, aggregations } = searchResults);
      const startHydrate = Date.now();
      results = await (new HydrateSerializer()).bulkSerialize(searchResults.output);
      hydrateDuration = Date.now() - startHydrate;
    }
    return {
      searchContent: results,
      resultCount,
      took: {
        total: Date.now() - start,
        hydrate: hydrateDuration,
        es: took,
      },
      aggregations,
    };
  }
}

const instance = new Searcher();
export default instance;
