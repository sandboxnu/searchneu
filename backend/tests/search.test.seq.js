import searcher from '../searcher';
import prisma from '../prisma';

beforeAll(async () => {
  searcher.subjects = [];
})

describe('searcher', () => {
  describe('generateMQuery', () => {
    it('generates with no filters', () => {
      expect(searcher.generateMQuery('fundies', '202030', 0, 10, {})).toMatchSnapshot();
    });

    it('generates aggs with online filters applied', () => {
      expect(searcher.generateMQuery('fundies', '202030', 0, 10, { online: true })).toMatchSnapshot();
    });
  });

  // TODO: create an association between cols in elasticCourseSerializer and here
  describe('generateQuery', () => {
    it('generates match_all when no query', () => {
      expect(searcher.generateQuery('', '202030', [], 0, 10).query.bool.must).toEqual({ match_all:{} });
    });

    it('generates a query without filters', () => {
      expect(searcher.generateQuery('fundies', '202030', [], 0, 10, 'nupath')).toMatchSnapshot();
    });
  });

  describe('validateFilters', () => {
    it('removes invalid filters', () => {
      const invalidFilters = {
        NUpath: 'NU Core/NUpath Adv Writ Dscpl',
        college: 'GS Col of Arts',
        subject: 'CS',
        online: false,
        classType: ['Lecture'],
        inValidFilterKey: '',
      };
      expect(searcher.validateFilters(invalidFilters)).toMatchObject({});
    });

    it('keeps all valid filters', () => {
      const validFilters = {
        nupath: ['NU Core/NUpath Adv Writ Dscpl', 'NUpath Interpreting Culture'],
        subject: ['ENGW', 'ARTG', 'CS'],
        online: true,
        classType: ['Lecture'],
      };
      expect(searcher.validateFilters(validFilters)).toMatchObject(validFilters);
    });
  });

  describe('Single search result', () => {
    beforeEach(async () => {
      await prisma.section.deleteMany({});
      await prisma.course.deleteMany({});
      await prisma.course.create({
        data: {
          id: 'neu.edu/202030/CS/2500',
          host: 'neu.edu',
          classId: '2500',
          name: 'Fundamentals of Computer Science 1',
          termId: '202030',
          subject: 'CS',
          lastUpdateTime: new Date(),
        },
      });
      await prisma.course.create({
        data: {
          id: 'neu.edu/202030/PHIL/1145',
          host: 'neu.edu',
          classId: '1145',
          name: 'Tech and Human Values',
          termId: '202030',
          subject: 'PHIL',
          nupath: { set: ['Ethical reasoning', 'Argue', 'Live in the mud'] },
        },
      });
      await prisma.course.create({
        data: {
          id: 'neu.edu/202030/CS/2510',
          host: 'neu.edu',
          classId: '2510',
          name: 'Fundamentals of Computer Science 2',
          termId: '202030',
          subject: 'CS',
        },
      });
      await prisma.section.create({
        data: {
          id: 'neu.edu/202030/CS/2500/19350',
          course: { connect: { id: 'neu.edu/202030/CS/2500' } },
          seatsCapacity: 80,
          seatsRemaining: 0,
          classType: 'Lecture',
        },
      });
      await prisma.section.create({
        data: {
          id: 'neu.edu/202030/PHIL/1145/20142',
          course: { connect: { id: 'neu.edu/202030/PHIL/1145' } },
          seatsCapacity: 40,
          seatsRemaining: 0,
          classType: 'Lecture',
        },
      });
    });
    describe('getSingleResultAggs', () => {
      it('Gets aggregation for single result', async () => {
        const singleResult = await prisma.course.findOne({ where: { id: 'neu.edu/202030/CS/2500' }, include: { sections: true } });
        expect(searcher.getSingleResultAggs(singleResult)).toEqual({
          nupath: [],
          subject: [{ value: 'CS', count: 1 }],
          classType: [{ value: 'Lecture', count: 1 }],
        });
      });
      it('Gets aggregation for single result with nupath', async () => {
        const singleResult = await prisma.course.findOne({ where: { id: 'neu.edu/202030/PHIL/1145' }, include: { sections: true } });
        expect(searcher.getSingleResultAggs(singleResult)).toEqual({
          nupath: [{ value: 'Ethical reasoning', count: 1 }, { value: 'Argue', count: 1 }, { value: 'Live in the mud', count: 1 }],
          subject: [{ value: 'PHIL', count: 1 }],
          classType: [{ value: 'Lecture', count: 1 }],
        });
      });
    });
    describe('getOneSearchResult', () => {
      it('Gets 1 result for valid class', async () => {
        expect(await searcher.getOneSearchResult('CS', '2500', '202030')).toMatchObject({
          results: [{
            class: {
              id: 'neu.edu/202030/CS/2500',
              host: 'neu.edu',
              classId: '2500',
              name: 'Fundamentals of Computer Science 1',
              termId: '202030',
              subject: 'CS',
              lastUpdateTime: expect.anything(),
            },
          }],
          resultCount: 1,
          took: 0,
          hydrateDuration: expect.anything(),
          aggregations: {
            nupath: [],
            subject: [{ value: 'CS', count: 1 }],
            classType: [{ value: 'Lecture', count: 1 }],
          },
        });
      });
      it('Gets 0 results for invalid course', async () => {
        expect(await searcher.getOneSearchResult('CS', '2504', '202030')).toMatchObject({
          results: [],
          resultCount: 0,
          took: 0,
          hydrateDuration: expect.anything(),
          aggregations: {
            nupath: [],
            subject: [],
            classType: [],
          },
        })
      });
    });
  })
});
