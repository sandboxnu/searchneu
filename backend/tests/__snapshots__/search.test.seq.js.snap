// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`searcher generateMQuery generates aggs with online filters applied 1`] = `
Array [
  Object {
    "aggregations": undefined,
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "sections.online": true,
                      },
                    },
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 10,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  Object {
    "aggregations": Object {
      "nupath": Object {
        "terms": Object {
          "field": "class.nupath.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "sections.online": true,
                      },
                    },
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  Object {
    "aggregations": Object {
      "subject": Object {
        "terms": Object {
          "field": "class.subject.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "sections.online": true,
                      },
                    },
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  Object {
    "aggregations": Object {
      "classType": Object {
        "terms": Object {
          "field": "sections.classType.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "term": Object {
                        "sections.online": true,
                      },
                    },
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
]
`;

exports[`searcher generateMQuery generates with no filters 1`] = `
Array [
  Object {
    "aggregations": undefined,
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              Object {
                "term": Object {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 10,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  Object {
    "aggregations": Object {
      "nupath": Object {
        "terms": Object {
          "field": "class.nupath.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              Object {
                "term": Object {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  Object {
    "aggregations": Object {
      "subject": Object {
        "terms": Object {
          "field": "class.subject.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              Object {
                "term": Object {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
  Object {
    "aggregations": Object {
      "classType": Object {
        "terms": Object {
          "field": "sections.classType.keyword",
          "size": 1000,
        },
      },
    },
    "from": 0,
    "query": Object {
      "bool": Object {
        "filter": Object {
          "bool": Object {
            "should": Array [
              Object {
                "bool": Object {
                  "must": Array [
                    Object {
                      "exists": Object {
                        "field": "sections",
                      },
                    },
                    Object {
                      "term": Object {
                        "class.termId": "202030",
                      },
                    },
                  ],
                },
              },
              Object {
                "term": Object {
                  "type": "employee",
                },
              },
            ],
          },
        },
        "must": Object {
          "multi_match": Object {
            "fields": Array [
              "class.name^2",
              "class.name.autocomplete",
              "class.subject^4",
              "class.classId^3",
              "sections.profs",
              "sections.crn",
              "employee.name^2",
              "employee.emails",
              "employee.phone",
            ],
            "query": "fundies",
            "type": "most_fields",
          },
        },
      },
    },
    "size": 0,
    "sort": Array [
      "_score",
      Object {
        "class.classId.keyword": Object {
          "order": "asc",
          "unmapped_type": "keyword",
        },
      },
    ],
  },
]
`;

exports[`searcher generateQuery generates a query without filters 1`] = `
Object {
  "aggregations": Object {
    "nupath": Object {
      "terms": Object {
        "field": "class.nupath.keyword",
        "size": 1000,
      },
    },
  },
  "from": 0,
  "query": Object {
    "bool": Object {
      "filter": Object {
        "bool": Object {
          "should": Array [
            Object {
              "bool": Object {
                "must": Array [
                  Object {
                    "exists": Object {
                      "field": "sections",
                    },
                  },
                  Object {
                    "term": Object {
                      "class.termId": "202030",
                    },
                  },
                ],
              },
            },
            Object {
              "term": Object {
                "type": "employee",
              },
            },
          ],
        },
      },
      "must": Object {
        "multi_match": Object {
          "fields": Array [
            "class.name^2",
            "class.name.autocomplete",
            "class.subject^4",
            "class.classId^3",
            "sections.profs",
            "sections.crn",
            "employee.name^2",
            "employee.emails",
            "employee.phone",
          ],
          "query": "fundies",
          "type": "most_fields",
        },
      },
    },
  },
  "size": 10,
  "sort": Array [
    "_score",
    Object {
      "class.classId.keyword": Object {
        "order": "asc",
        "unmapped_type": "keyword",
      },
    },
  ],
}
`;
